### ISSUE TITLE
[STORM-160] Allow ShellBolt to set env vars (particularly PATH)

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#4002]
<p><a href="https://github.com/nathanmarz/storm/issues/32" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/nathanmarz/storm/issues/32</a></p>

<p>While trying to implement a multilang based bolt, I discovered that binaries outside of /usr/bin, /usr/sbin, and /sbin, are not found despite being installed in say, /usr/local/bin.</p>

<p>Is there a way to specify a PATH variable so when storm creates the sub shell, these binaries can be located by name? If not, can we get one added?</p>

<p>----------<br/>
nathanmarz: This is a good idea. For the meantime, does using the full path of the binary (e.g., /usr/local/bin/python) work around this issue?</p>

<p>----------<br/>
dinedal: That works for processes that don't require any other environment variables.</p>

<p>Also, in cases where developers have multiple versions of a program (say, RVM) and the production servers have another one installed on the system, it means changing code for deploying the topology.</p>

<p>----------<br/>
nicoo: +1</p>

<p>To work around this issue you can start your multilang bolt from a bash script that set all needed env vars.</p>

---
<details><summary><i>Originally reported by <a title="James Xu" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=xumingming">xumingming</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-160" target="_blank">Allow ShellBolt to set env vars (particularly PATH)</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Shyam Rajendran" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran">shyamrajendran</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-07-13T22:09:43+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-109074487" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-109074487</a></p>

<p>    Regarding test failures, there're random failures on Kafka tests with Travis CI. <br/>
    So if failure occurs only Kafka tests, you can ignore them for now.</p>

<p>    LGTM on changeset.<br/>
    Btw, seems like we can apply this to ShellSpout, too. What do you think?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-109361408" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-109361408</a></p>

<p>    Thanks HeartSaVioR. Yes we can apply this to ShellSpout as well. I will do the changes shortly. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-110366648" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-110366648</a></p>

<p>    The changes to ShellSpout have been committed now. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-110522545" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-110522545</a></p>

<p>    LGTM.<br/>
    Btw, there're many unrelated commits so I can hardly see actual changeset easily.<br/>
    How about creating new pull request or rebase which points to current master + actual changeset?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-111151124" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-111151124</a></p>

<p>    Thanks HeartSavior. I have cleaned up the commit history now. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#discussion_r33607552" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#discussion_r33607552</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/backtype/storm/utils/ShellProcess.java &#8212;<br/>
    @@ -18,16 +18,14 @@<br/>
     package backtype.storm.utils;</p>

<p>     import backtype.storm.Config;<br/>
    -import backtype.storm.multilang.ISerializer;<br/>
    -import backtype.storm.multilang.BoltMsg;<br/>
    -import backtype.storm.multilang.NoOutputException;<br/>
    -import backtype.storm.multilang.ShellMsg;<br/>
    -import backtype.storm.multilang.SpoutMsg;<br/>
    +import backtype.storm.multilang.*;<br/>
    &#8212; End diff &#8211;</p>

<p>    Let's leave these expanded.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#discussion_r33607948" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#discussion_r33607948</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/backtype/storm/utils/ShellProcess.java &#8212;<br/>
    @@ -48,8 +47,26 @@ public ShellProcess(String[] command) </p>
{
     this.command = command;
 }

<p>    +    public void setEnv(Map<String, String> env) </p>
{
    +this.env = env;
    +    }
<p>    +<br/>
    +    private void modifyEnvironment(Map<String, String> buildEnv) {<br/>
    +for (Map.Entry<String, String> entry : env.entrySet()) {<br/>
    +    if (entry.getKey().equals("PATH")) {<br/>
    &#8212; End diff &#8211;</p>

<p>    Just on the safe side: Let's use the form `"PATH".equals ...` in case the key is null for whatever reason.</p>

<p>    I think with a Map this is impossible, but should there be a code change later, we can trivially avoid an NPE here.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#discussion_r33608225" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#discussion_r33608225</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/backtype/storm/utils/ShellProcess.java &#8212;<br/>
    @@ -48,8 +47,26 @@ public ShellProcess(String[] command) </p>
{
     this.command = command;
 }

<p>    +    public void setEnv(Map<String, String> env) </p>
{
    +this.env = env;
    +    }
<p>    +<br/>
    +    private void modifyEnvironment(Map<String, String> buildEnv) {<br/>
    +for (Map.Entry<String, String> entry : env.entrySet()) {<br/>
    +    if (entry.getKey().equals("PATH")) {<br/>
    +buildEnv.put("PATH", buildEnv.get("PATH") + ":" + env.get("PATH"));<br/>
    &#8212; End diff &#8211;</p>

<p>    Let's use the file-system generic path separator here. File#pathSeparator</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-117298122" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-117298122</a></p>

<p>    Overall looks good to me.  Just a few very minor comments.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-117777990" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-117777990</a></p>

<p>    Thanks Derek for the comments. I have now pushed in the changes.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-117865734" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-117865734</a></p>

<p>    @bourneagain <br/>
    Travis CI complains that your PR has compile failure cause of missing imports. Could you fix it?</p>

<p>    One minor note, please post / add commits after building / testing from your local machine. Thanks!</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-118387380" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-118387380</a></p>

<p>    Thanks @HeartSaVioR . I have fixed the imports issue and tested on local machine to see it working fine. I shall keep a note of this in future. Thanks.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#discussion_r33884760" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#discussion_r33884760</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/backtype/storm/utils/ShellProcess.java &#8212;<br/>
    @@ -48,8 +50,31 @@ public ShellProcess(String[] command) </p>
{
     this.command = command;
 }

<p>    +    public void setEnv(Map<String, String> env) </p>
{
    +this.env = env;
    +    }
<p>    +<br/>
    +    private void modifyEnvironment(Map<String, String> buildEnv) {<br/>
    +for (Map.Entry<String, String> entry : env.entrySet()) {<br/>
    +    if ("PATH".equals(entry.getKey())) </p>
{
    +buildEnv.put("PATH", buildEnv.get("PATH") + File.pathSeparatorChar + env.get("PATH"));
    +    }
<p> else </p>
{
    +buildEnv.put(entry.getKey(), entry.getValue());
    +    }
<p>    +}<br/>
    +    }<br/>
    +<br/>
    +<br/>
 public Number launch(Map conf, TopologyContext context) {<br/>
     ProcessBuilder builder = new ProcessBuilder(command);<br/>
    +if (!env.isEmpty()) </p>
{
    +    Map<String, String> buildEnv = builder.environment();
    +    modifyEnvironment(buildEnv);
    +}
<p>    +for(Map.Entry<String, String> entry : builder.environment().entrySet()) {<br/>
    &#8212; End diff &#8211;</p>

<p>    @bourneagain <br/>
    Maybe it is for debugging. Could you remove it, or use logger to log properly?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#discussion_r33894593" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#discussion_r33894593</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/backtype/storm/utils/ShellProcess.java &#8212;<br/>
    @@ -48,8 +50,31 @@ public ShellProcess(String[] command) </p>
{
     this.command = command;
 }

<p>    +    public void setEnv(Map<String, String> env) </p>
{
    +this.env = env;
    +    }
<p>    +<br/>
    +    private void modifyEnvironment(Map<String, String> buildEnv) {<br/>
    +for (Map.Entry<String, String> entry : env.entrySet()) {<br/>
    +    if ("PATH".equals(entry.getKey())) </p>
{
    +buildEnv.put("PATH", buildEnv.get("PATH") + File.pathSeparatorChar + env.get("PATH"));
    +    }
<p> else </p>
{
    +buildEnv.put(entry.getKey(), entry.getValue());
    +    }
<p>    +}<br/>
    +    }<br/>
    +<br/>
    +<br/>
 public Number launch(Map conf, TopologyContext context) {<br/>
     ProcessBuilder builder = new ProcessBuilder(command);<br/>
    +if (!env.isEmpty()) </p>
{
    +    Map<String, String> buildEnv = builder.environment();
    +    modifyEnvironment(buildEnv);
    +}
<p>    +for(Map.Entry<String, String> entry : builder.environment().entrySet()) {<br/>
    &#8212; End diff &#8211;</p>

<p>    @HeartSaVioR , Sorry I am not sure what you meant here. The code (line 70) is checking if the user had passed any custom environment map only after which it calls the modifyEnvironment with the existing environment map for modification.  Dd you mean to include debug message for better clarity within this block ?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#discussion_r33895557" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#discussion_r33895557</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/backtype/storm/utils/ShellProcess.java &#8212;<br/>
    @@ -48,8 +50,31 @@ public ShellProcess(String[] command) </p>
{
     this.command = command;
 }

<p>    +    public void setEnv(Map<String, String> env) </p>
{
    +this.env = env;
    +    }
<p>    +<br/>
    +    private void modifyEnvironment(Map<String, String> buildEnv) {<br/>
    +for (Map.Entry<String, String> entry : env.entrySet()) {<br/>
    +    if ("PATH".equals(entry.getKey())) </p>
{
    +buildEnv.put("PATH", buildEnv.get("PATH") + File.pathSeparatorChar + env.get("PATH"));
    +    }
<p> else </p>
{
    +buildEnv.put(entry.getKey(), entry.getValue());
    +    }
<p>    +}<br/>
    +    }<br/>
    +<br/>
    +<br/>
 public Number launch(Map conf, TopologyContext context) {<br/>
     ProcessBuilder builder = new ProcessBuilder(command);<br/>
    +if (!env.isEmpty()) </p>
{
    +    Map<String, String> buildEnv = builder.environment();
    +    modifyEnvironment(buildEnv);
    +}
<p>    +for(Map.Entry<String, String> entry : builder.environment().entrySet()) {<br/>
    &#8212; End diff &#8211;</p>

<p>    @bourneagain<br/>
    Github points out different line, I mean 74 line and belows, why call System.out.println() for each entry, prefix with "SHYAM:"?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-118633914" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-118633914</a></p>

<p>    Thanks @HeartSaVioR . I apologize for the mistake. I have now removed the local debug I had put on during testing. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-118785261" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-118785261</a></p>

<p>    +1, LGTM.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran">shyamrajendran</a>:</i>
<p>The user environment variable map can be now passed on during the building of the topology to either Spout or Bolt. PATH variable alone gets appended while other envs get overwritten if set by the user.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran">shyamrajendran</a>:</i>
<p>The patch requires one more +1 </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-120450816" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-120450816</a></p>

<p>    @HeartSaVioR  Thanks. Can this be merged into the master now?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-120932159" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-120932159</a></p>

<p>    @bourneagain Yes, let's merge. <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-120938352" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-120938352</a></p>

<p>    I merged it into master branch at ASF git repos, but seems like Github mirror is out of sync.</p>

<p>    I'll resolve issue when mirror keeps up ASF repo.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user bourneagain commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/575#issuecomment-120938865" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575#issuecomment-120938865</a></p>

<p>    Thank you @HeartSaVioR </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/575" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/575</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Github mirror catches up ASF now.<br/>
Thanks <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamrajendran" class="user-hover" rel="shyamrajendran">Shyam Rajendran</a> for great work.</p>
